src/providers/MeetingProvider/MeetingManager.ts

```diff
@@ -314,7 +329,19 @@ export class MeetingManager implements AudioVideoObserver {
 
   async listAndSelectDevices(): Promise<void> {
     await this.updateDeviceLists();
+
+    let hasAudioInput = false, hasAudioOutput = false, hasVideoInput = false;
+    const devices = await navigator.mediaDevices.enumerateDevices();
+    for (const device of devices) {
+      if (device.label) {
+        if (device.kind === "audioinput") hasAudioInput = true;
+        if (device.kind === "audiooutput") hasAudioOutput = true;
+        if (device.kind === "videoinput") hasVideoInput = true;
+      }
+    }
```

Comment:
This may be a breakthrough change. The main logic here is that we only call `choose*Device` for those devices with labels.

For example, set `deviceLabels` to `DeviceLabels.Video`. On the device setting page, the first device is selected for video input, and empty labels are displayed for audio input and audio output. Later on the meeting page, the first device is selected for video input, and "None" is selected for audio input and audio output.

Reply:
It is OK to choose devices that don't have labels…

Reply:
@richnew10 This is to handle the error and enhance user experience. 
In `listAndSelectDevices` function, we call `audioVideo?.choose*Device()` which may call `getUserMedia()`. This could lead to some error which users don't expect, after we add the device permission control feature.

For example, user wants to join in view-only mode. They pass the `deviceLabels.None` via `mm.join(deviceLabels)`. This will set the `deviceLabelTrigger` as `() => Promise.resolve(new MediaStream());`. And if we don't make this change, when user joins the meeting, there will be errors like below:
![image](https://user-images.githubusercontent.com/20471438/121099636-10335000-c7ad-11eb-9d86-430590cb90de.png)

Reply:
My point is: browsers encourage applications to persist device IDs across application sessions.

So you might need to get labels _once_, and after that you'll just call

```typescript
audioVideo.chooseAudioInputDevice('some-persisted-id');
```

without (a) that device having a label, (b) having called `enumerateDevices`, or (c) having triggered a permission request with a hook.

Such an application will use `DeviceLabels.None` to avoid triggering an arbitrary microphone and video input for no reason, and will then directly choose an input. At no point does that application want `enumerateDevices` being called or the device label hook to result in `getUserMedia` being called.